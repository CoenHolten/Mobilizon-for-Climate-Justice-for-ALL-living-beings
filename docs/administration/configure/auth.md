# Authentification

## LDAP

Use LDAP for user authentication.  When a user logs in to the Mobilizon instance, the email and password will be verified by trying to authenticate
(bind) to an LDAP server. If a user exists in the LDAP directory but there is no account with the same email yet on the Mobilizon instance then a new
Mobilizon account will be created (without needing email confirmation) with the same email as the LDAP email name.

!!! tip
    As Mobilizon uses email for login and LDAP bind is often done with account UID/CN, we need to start by searching for LDAP account matching with this email. LDAP search without bind is often disallowed, so you'll probably need an admin LDAP user.


Change authentification method:
```elixir
config :mobilizon,
       Mobilizon.Service.Auth.Authenticator,
       Mobilizon.Service.Auth.LDAPAuthenticator
```

LDAP configuration under `:mobilizon, :ldap`:

* `enabled`: enables LDAP authentication
* `host`: LDAP server hostname
* `port`: LDAP port, e.g. 389 or 636
* `ssl`: true to use SSL, usually implies the port 636
* `sslopts`: additional SSL options
* `tls`: true to start TLS, usually implies the port 389
* `tlsopts`: additional TLS options
* `base`: LDAP base, e.g. "dc=example,dc=com"
* `uid`: LDAP attribute name to authenticate the user, e.g. when "cn", the filter will be "cn=username,base"
* `require_bind_for_search` whether admin bind is required to perform search
* `bind_uid` the admin uid/cn for binding before searching
* `bind_password` the admin password for binding before searching

Example:

```elixir
config :mobilizon, :ldap,
  enabled: true,
  host: "localhost",
  port: 636,
  ssl: true,
  sslopts: [],
  tls: true,
  tlsopts: [],
  base: "ou=users,dc=example,dc=local",
  uid: "cn",
  require_bind_for_search: true,
  bind_uid: "admin_account",
  bind_password: "some_admin_password"
```

## OAuth

Mobilizon currently supports the following providers:

* [Discord](https://github.com/schwarz/ueberauth_discord)
* [Facebook](https://github.com/ueberauth/ueberauth_facebook)
* [Github](https://github.com/ueberauth/ueberauth_github)
* [Gitlab](https://github.com/mtchavez/ueberauth_gitlab) (including self-hosted)
* [Google](https://github.com/ueberauth/ueberauth_google)
* [Keycloak](https://github.com/Rukenshia/ueberauth_keycloak) (through OpenID Connect)
* [Twitter](https://github.com/Rukenshia/ueberauth_keycloak)

Support for [other providers](https://github.com/ueberauth/ueberauth/wiki/List-of-Strategies) can easily be added if requested.

!!! tip
    We advise to look at each provider's README file for eventual specific instructions.

You'll have to start by registering an app at the provider. Be sure to activate features like "Sign-in with" and "emails" scope, as Mobilizon needs users emails to register them.

Add the configured providers to configuration (you may find the appropriate scopes on the provider's API documentation):
```elixir
config :ueberauth,
       Ueberauth,
       providers: [
         gitlab: {Ueberauth.Strategy.Gitlab, [default_scope: "read_user"]},
         keycloak: {Ueberauth.Strategy.Keycloak, [default_scope: "email"]}
         # ...
       ]
```

In order for the « Sign-in with » buttons to be added on Register and Login pages, list your providers:
```elixir
config :mobilizon, :auth,
  oauth_consumer_strategies: [
    :gitlab,
    {:keycloak, "My corporate account"}
    # ...
  ]
```

!!! note
    If you use the `{:provider_id, "Some label"}` form, the label will be used inside the buttons on Register and Login pages.

Finally add the configuration for each specific provider. The Client ID and Client Secret are at least required:
```elixir
config :ueberauth, Ueberauth.Strategy.Facebook.OAuth,
  client_id: "some_numeric_id",
  client_secret: "some_secret"

keycloak_url = "https://some-keycloak-instance.org"

# Realm may be something else than master
config :ueberauth, Ueberauth.Strategy.Keycloak.OAuth,
  client_id: "some_id",
  client_secret: "some_hexadecimal_secret",
  site: keycloak_url,
  authorize_url: "#{keycloak_url}/auth/realms/master/protocol/openid-connect/auth",
  token_url: "#{keycloak_url}/auth/realms/master/protocol/openid-connect/token",
  userinfo_url: "#{keycloak_url}/auth/realms/master/protocol/openid-connect/userinfo",
  token_method: :post
```