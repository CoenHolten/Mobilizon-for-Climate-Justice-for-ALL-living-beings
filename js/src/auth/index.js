import router from '../router/index';
import { API_HOST, API_PATH } from '../api/_entrypoint';

// URL and endpoint constants
const LOGIN_URL = `${API_HOST}${API_PATH}/login`;
const SIGNUP_URL = `${API_HOST}${API_PATH}/users/`;
const CHECK_AUTH = `${API_HOST}${API_PATH}/users/`;
const REFRESH_TOKEN = `${API_HOST}${API_PATH}/token/refresh`;

function AuthError(field, message) {
  this.field = field;
  this.message = message;
}

AuthError.prototype.toString = function AuthErrorToString() {
  return `AuthError: ${this.message}`;
};

export default {

  // User object will let us check authentication status
  user: false,
  authenticated: false,
  token: false,

  // Send a request to the login URL and save the returned JWT
  login(creds, $store, redirect, error) {
    fetch(LOGIN_URL, { method: 'POST', body: creds, headers: { 'Content-Type': 'application/json' } })
      .then(response => response.json())
      .then((data) => {
        if (data.code >= 300) {
          throw new AuthError(null, data.message);
        }
        $store.commit('LOGIN_USER');

        localStorage.setItem('token', data.token);
        localStorage.setItem('refresh_token', data.refresh_token);
        this.getUser(
          $store,
          () => router.push(redirect)
        );

      }).catch((err) => {
        error(err);
      });
  },

  signup(creds, $store, redirect, error) {
    fetch(SIGNUP_URL, { method: 'POST', body: creds, headers: { 'Content-Type': 'application/json' } })
      .then(response => response.json())
      .then((data) => {
        if (data.error) {
          throw new AuthError(data.error.field, data.error.message);
        }

        $store.commit('LOGIN_USER');
        localStorage.setItem('token', data.token);
        localStorage.setItem('refresh_token', data.refresh_token);

        if (redirect) {
          router.push(redirect);
        }
      }).catch((err) => {
        error(err);
      });
  },
  refreshToken(store, successHandler, errorHandler) {
    const refreshToken = localStorage.getItem('refresh_token');
    console.log("We are refreshing the jwt token");
    fetch(REFRESH_TOKEN, { method: 'POST', body: JSON.stringify({refresh_token: refreshToken}), headers: { 'Content-Type': 'application/json' }})
      .then((response) => {
        if (response.ok) {
          return response.json();
        } else {
          errorHandler('Error while authenticating');
        }
      })
      .then((response) => {
        console.log("We have a new token");
        this.authenticated = true;
        store.commit('LOGIN_USER', response);
        localStorage.setItem('token', response.token);
        console.log("Let's try to auth again");
        this.getUser(store, successHandler, errorHandler);
        successHandler();
      });
  },

  // To log out, we just need to remove the token
  logout() {
    localStorage.removeItem('refresh_token');
    localStorage.removeItem('token');
    this.authenticated = false;
  },

  jwt_decode(token) {
    const base64Url = token.split('.')[1];
    const base64 = base64Url.replace('-', '+').replace('_', '/');
    return JSON.parse(window.atob(base64));
  },

  getTokenExpirationDate(encodedToken) {
    const token = this.jwt_decode(encodedToken);
    if (!token.exp) { return null; }

    const date = new Date(0);
    date.setUTCSeconds(token.exp);

    return date;
  },

  isTokenExpired(token) {
    const expirationDate = this.getTokenExpirationDate(token);
    return expirationDate < new Date();
  },

  checkAuth(store = null) {
    const token = localStorage.getItem('token');
    if (store && token) {
      this.getUser(store,() => null, () => null);
    }
    /* if (!!token && store && !this.isTokenExpired(token)) {
      this.refreshToken(store, () => null, () => null);
    } */
    return !!token;
  },

  getUser(store, successHandler, errorHandler) {
    console.log("We are checking the auth");
    this.token = localStorage.getItem('token');
    const options = {};
    options.headers = new Headers();
    options.headers.set('Authorization', `Bearer ${this.token}`);
    fetch(CHECK_AUTH, options)
      .then((response) => {
        if (response.ok) {
          return response.json();
        } else {
          errorHandler('Error while authenticating');
        }
      })
      .then((response) => {
        this.authenticated = true;
        console.log(response);
        store.commit('SAVE_USER', response);
        successHandler();
      });
  },

  // The object to be passed as a header for authenticated requests
  getAuthHeader() {
    return {
      Authorization: `Bearer ${localStorage.getItem('access_token')}`,
    };
  },
};
