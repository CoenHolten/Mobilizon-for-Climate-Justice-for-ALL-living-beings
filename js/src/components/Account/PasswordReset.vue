<template>
  <v-container fluid fill-height>
    <v-layout align-center justify-center>
      <v-flex xs12 sm8 md4>
        <v-card class="elevation-12">
          <v-toolbar dark color="primary">
            <v-toolbar-title>Password Reset</v-toolbar-title>
          </v-toolbar>
          <v-card-text>
            <v-alert type="error" :value="state.token.status === false">{{ state.token.msg }}</v-alert>
            <v-form @submit="resetAction">
              <v-text-field
                label="Password"
                type="password"
                v-model="credentials.password"
                required
                :error="state.password.status"
                :rules="[rules.required, rules.password_length]"
              >
              </v-text-field>
              <v-text-field
                label="Password (confirmation)"
                type="password"
                v-model="credentials.password_confirmation"
                required
                :rules="[rules.required, rules.password_length, rules.password_equal]"
                :error="state.password_confirmation.status"
              >
              </v-text-field>
              <v-btn type="submit" :disabled="!samePasswords" color="blue">Reset my password</v-btn>
            </v-form>
          </v-card-text>
        </v-card>
      </v-flex>
    </v-layout>
  </v-container>
</template>

<script lang="ts">

  import { Component, Prop, Vue } from 'vue-property-decorator';

  @Component
  export default class PasswordReset extends Vue {
    @Prop({ type: String, required: true }) token!: string;

    credentials = {
      password: '',
      password_confirmation: '',
    };
    error = {
      show: false,
    };
    state = {
      token: {
        status: null,
        msg: '',
      },
      password: {
        status: null,
        msg: '',
      },
      password_confirmation: {
        status: null,
        msg: '',
      },
    };
    rules = {
      password_length: value => value.length > 6 || 'Password must be at least 6 caracters long',
      required: value => !!value || 'Required.',
      password_equal: value => value === this.credentials.password || 'Passwords must be the same',
    };

    get samePasswords() {
      return this.rules.password_length(this.credentials.password) === true &&
        this.credentials.password === this.credentials.password_confirmation;
    }

    resetAction(e) {
      this.resetState();
      e.preventDefault();
      console.log(this.token);
      // FIXME: implements fetchStory
      // fetchStory('/users/password-reset/post', this.$store, {
      //   method: 'POST',
      //   body: JSON.stringify({ password: this.credentials.password, token: this.token }),
      // }).then((data) => {
      //   localStorage.setItem('token', data.token);
      //   localStorage.setItem('refresh_token', data.refresh_token);
      //   this.$store.commit('LOGIN_USER', data.account);
      //   this.$snotify.success(this.$t('registration.success.login', { username: data.account.username }));
      //   this.$router.push({ name: 'Home' });
      // }, (error) => {
      //   Promise.resolve(error).then((errormsg) => {
      //     console.log('errormsg', errormsg);
      //     this.error.show = true;
      //     Object.entries(JSON.parse(errormsg).errors).forEach(([ key, val ]) => {
      //       console.log('key', key);
      //       console.log('val', val[ 0 ]);
      //       this.state[ key ] = { status: false, msg: val[ 0 ] };
      //       console.log('state', this.state);
      //     });
      //   });
      // });
    }

    resetState() {
      this.state = {
        token: {
          status: null,
          msg: '',
        },
        password_confirmation: {
          status: null,
          msg: '',
        },
        password: {
          status: null,
          msg: '',
        },
      };
    }
  };
</script>
