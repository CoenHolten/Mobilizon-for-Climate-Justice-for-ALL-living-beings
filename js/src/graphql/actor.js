import gql from 'graphql-tag';

export const FETCH_ACTOR = gql`
query($name:String!) {
  actor(preferredUsername: $name) {
    url,
    outboxUrl,
    inboxUrl,
    followingUrl,
    followersUrl,
    sharedInboxUrl,
    name,
    domain,
    summary,
    preferredUsername,
    suspended,
    avatarUrl,
    bannerUrl,
    organizedEvents {
        uuid,
        title,
        description,
        organizer_actor {
            avatarUrl,
            preferred_username,
            name,
        }
    },
  }
}
`;

export const LOGGED_ACTOR = gql`
query {
  loggedActor {
    avatarUrl,
    preferredUsername,
  }
}`;
