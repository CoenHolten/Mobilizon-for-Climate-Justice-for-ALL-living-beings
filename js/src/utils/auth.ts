import { AUTH_TOKEN, AUTH_USER_EMAIL, AUTH_USER_ID } from '@/constants';
import { ILogin } from '@/types/login.model';

export function saveUserData(obj: ILogin) {
  localStorage.setItem(AUTH_USER_ID, `${obj.user.id}`);
  localStorage.setItem(AUTH_USER_EMAIL, obj.user.email);
  localStorage.setItem(AUTH_TOKEN, obj.token);
}

export function deleteUserData() {
  for (const key of [AUTH_USER_ID, AUTH_USER_EMAIL, AUTH_TOKEN]) {
    localStorage.removeItem(key);
  }
}
