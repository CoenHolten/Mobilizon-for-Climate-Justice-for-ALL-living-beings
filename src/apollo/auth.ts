import { AUTH_ACCESS_TOKEN } from "@/constants";
import { ApolloLink } from "@apollo/client/core";

export function generateTokenHeader() {
  const token = localStorage.getItem(AUTH_ACCESS_TOKEN);

  return token ? `Bearer ${token}` : null;
}

const authMiddleware = new ApolloLink((operation, forward) => {
  // add the authorization to the headers
  operation.setContext({
    headers: {
      authorization: generateTokenHeader(),
    },
  });

  if (forward) return forward(operation);

  return null;
});

export { authMiddleware };
