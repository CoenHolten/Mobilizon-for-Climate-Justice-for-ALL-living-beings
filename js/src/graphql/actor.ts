import gql from "graphql-tag";
import { DISCUSSION_BASIC_FIELDS_FRAGMENT } from "@/graphql/discussion";
import { RESOURCE_METADATA_BASIC_FIELDS_FRAGMENT } from "@/graphql/resources";
import { POST_BASIC_FIELDS } from "./post";

export const FETCH_PERSON = gql`
  query($username: String!) {
    fetchPerson(preferredUsername: $username) {
      id
      url
      name
      domain
      summary
      preferredUsername
      suspended
      avatar {
        name
        url
      }
      banner {
        url
      }
      feedTokens {
        token
      }
      organizedEvents {
        total
        elements {
          uuid
          title
          beginsOn
        }
      }
    }
  }
`;

export const GET_PERSON = gql`
  query(
    $actorId: ID!
    $organizedEventsPage: Int
    $organizedEventsLimit: Int
    $participationPage: Int
    $participationLimit: Int
  ) {
    person(id: $actorId) {
      id
      url
      name
      domain
      summary
      preferredUsername
      suspended
      avatar {
        name
        url
      }
      banner {
        url
      }
      feedTokens {
        token
      }
      organizedEvents(page: $organizedEventsPage, limit: $organizedEventsLimit) {
        total
        elements {
          id
          uuid
          title
          beginsOn
        }
      }
      participations(page: $participationPage, limit: $participationLimit) {
        total
        elements {
          id
          event {
            id
            uuid
            title
            beginsOn
          }
        }
      }
      user {
        id
        email
      }
    }
  }
`;

export const LIST_PROFILES = gql`
  query ListProfiles(
    $preferredUsername: String
    $name: String
    $domain: String
    $local: Boolean
    $suspended: Boolean
    $page: Int
    $limit: Int
  ) {
    persons(
      preferredUsername: $preferredUsername
      name: $name
      domain: $domain
      local: $local
      suspended: $suspended
      page: $page
      limit: $limit
    ) {
      total
      elements {
        id
        preferredUsername
        domain
        name
        avatar {
          url
        }
      }
    }
  }
`;

export const LOGGED_PERSON = gql`
  query {
    loggedPerson {
      id
      avatar {
        url
      }
      preferredUsername
    }
  }
`;

export const UPDATE_DEFAULT_ACTOR = gql`
  mutation ChangeDefaultActor($preferredUsername: String!) {
    changeDefaultActor(preferredUsername: $preferredUsername) {
      id
      defaultActor {
        id
      }
    }
  }
`;

export const CURRENT_ACTOR_CLIENT = gql`
  query currentActor {
    currentActor @client {
      id
      avatar {
        url
      }
      preferredUsername
      name
    }
  }
`;

export const UPDATE_CURRENT_ACTOR_CLIENT = gql`
  mutation UpdateCurrentActor(
    $id: String!
    $avatar: String
    $preferredUsername: String!
    $name: String!
  ) {
    updateCurrentActor(
      id: $id
      avatar: $avatar
      preferredUsername: $preferredUsername
      name: $name
    ) @client
  }
`;

export const LOGGED_USER_PARTICIPATIONS = gql`
  query LoggedUserParticipations(
    $afterDateTime: DateTime
    $beforeDateTime: DateTime
    $page: Int
    $limit: Int
  ) {
    loggedUser {
      participations(
        afterDatetime: $afterDateTime
        beforeDatetime: $beforeDateTime
        page: $page
        limit: $limit
      ) {
        total
        elements {
          event {
            id
            uuid
            title
            picture {
              id
              url
              alt
            }
            beginsOn
            visibility
            organizerActor {
              id
              preferredUsername
              name
              domain
              summary
              avatar {
                url
              }
            }
            participantStats {
              going
              notApproved
              participant
            }
            options {
              maximumAttendeeCapacity
              remainingAttendeeCapacity
            }
          }
          id
          role
          actor {
            id
            preferredUsername
            name
            domain
            summary
            avatar {
              url
            }
          }
        }
      }
    }
  }
`;

export const LOGGED_USER_DRAFTS = gql`
  query LoggedUserDrafts($page: Int, $limit: Int) {
    loggedUser {
      drafts(page: $page, limit: $limit) {
        id
        uuid
        title
        picture {
          id
          url
          alt
        }
        beginsOn
        visibility
        organizerActor {
          id
          preferredUsername
          name
          domain
          avatar {
            url
          }
        }
        participantStats {
          going
          notApproved
        }
        options {
          maximumAttendeeCapacity
          remainingAttendeeCapacity
        }
      }
    }
  }
`;

export const LOGGED_USER_MEMBERSHIPS = gql`
  query LoggedUserMemberships($page: Int, $limit: Int) {
    loggedUser {
      memberships(page: $page, limit: $limit) {
        total
        elements {
          id
          role
          parent {
            id
            preferredUsername
            domain
            name
            avatar {
              url
            }
            organizedEvents {
              elements {
                id
                title
                picture {
                  url
                }
              }
              total
            }
          }
          invitedBy {
            id
            preferredUsername
            domain
            name
            avatar {
              url
            }
          }
        }
      }
    }
  }
`;

export const IDENTITIES = gql`
  query {
    identities {
      id
      avatar {
        url
      }
      preferredUsername
      name
    }
  }
`;

export const PERSON_MEMBERSHIPS = gql`
  query PersonMemberships($id: ID!) {
    person(id: $id) {
      id
      memberships {
        total
        elements {
          id
          role
          parent {
            id
            preferredUsername
            name
            domain
            avatar {
              url
            }
          }
        }
      }
    }
  }
`;

export const CREATE_PERSON = gql`
  mutation CreatePerson(
    $preferredUsername: String!
    $name: String!
    $summary: String
    $avatar: PictureInput
  ) {
    createPerson(
      preferredUsername: $preferredUsername
      name: $name
      summary: $summary
      avatar: $avatar
    ) {
      id
      preferredUsername
      name
      summary
      avatar {
        url
      }
    }
  }
`;

export const UPDATE_PERSON = gql`
  mutation UpdatePerson($id: ID!, $name: String, $summary: String, $avatar: PictureInput) {
    updatePerson(id: $id, name: $name, summary: $summary, avatar: $avatar) {
      id
      preferredUsername
      name
      summary
      avatar {
        url
      }
    }
  }
`;

export const DELETE_PERSON = gql`
  mutation DeletePerson($id: ID!) {
    deletePerson(id: $id) {
      preferredUsername
    }
  }
`;

/**
 * This one is used only to register the first account.
 * Prefer CREATE_PERSON when creating another identity
 */
export const REGISTER_PERSON = gql`
  mutation($preferredUsername: String!, $name: String!, $summary: String!, $email: String!) {
    registerPerson(
      preferredUsername: $preferredUsername
      name: $name
      summary: $summary
      email: $email
    ) {
      id
      preferredUsername
      name
      summary
      avatar {
        url
      }
    }
  }
`;

export const LIST_GROUPS = gql`
  query {
    groups {
      elements {
        id
        url
        name
        domain
        summary
        preferredUsername
        suspended
        avatar {
          url
        }
        banner {
          url
        }
        organizedEvents {
          elements {
            uuid
            title
            beginsOn
          }
          total
        }
      }
      total
    }
  }
`;

export const FETCH_GROUP = gql`
  query($name: String!) {
    group(preferredUsername: $name) {
      id
      url
      name
      domain
      summary
      preferredUsername
      suspended
      visibility
      physicalAddress {
        description
        street
        locality
        postalCode
        region
        country
        geom
        type
        id
        originId
      }
      avatar {
        url
      }
      banner {
        url
      }
      organizedEvents {
        elements {
          id
          uuid
          title
          beginsOn
        }
        total
      }
      discussions {
        total
        elements {
          ...DiscussionBasicFields
        }
      }
      posts {
        total
        elements {
          ...PostBasicFields
        }
      }
      members {
        elements {
          role
          actor {
            id
            name
            domain
            preferredUsername
            avatar {
              url
            }
          }
          insertedAt
        }
        total
      }
      resources(page: 1, limit: 3) {
        elements {
          id
          title
          resourceUrl
          summary
          updatedAt
          type
          path
          metadata {
            ...ResourceMetadataBasicFields
          }
        }
        total
      }
      todoLists {
        elements {
          id
          title
          todos {
            elements {
              id
              title
              status
              dueDate
              assignedTo {
                id
                preferredUsername
              }
            }
            total
          }
        }
        total
      }
    }
  }
  ${DISCUSSION_BASIC_FIELDS_FRAGMENT}
  ${POST_BASIC_FIELDS}
  ${RESOURCE_METADATA_BASIC_FIELDS_FRAGMENT}
`;

export const CREATE_GROUP = gql`
  mutation CreateGroup(
    $creatorActorId: ID!
    $preferredUsername: String!
    $name: String!
    $summary: String
    $avatar: PictureInput
    $banner: PictureInput
  ) {
    createGroup(
      creatorActorId: $creatorActorId
      preferredUsername: $preferredUsername
      name: $name
      summary: $summary
      banner: $banner
      avatar: $avatar
    ) {
      id
      preferredUsername
      name
      summary
      avatar {
        url
      }
      banner {
        url
      }
    }
  }
`;

export const UPDATE_GROUP = gql`
  mutation UpdateGroup(
    $id: ID!
    $name: String
    $summary: String
    $avatar: PictureInput
    $banner: PictureInput
    $visibility: GroupVisibility
    $physicalAddress: AddressInput
  ) {
    updateGroup(
      id: $id
      name: $name
      summary: $summary
      banner: $banner
      avatar: $avatar
      visibility: $visibility
      physicalAddress: $physicalAddress
    ) {
      id
      preferredUsername
      name
      summary
      avatar {
        url
      }
      banner {
        url
      }
    }
  }
`;

export const SUSPEND_PROFILE = gql`
  mutation SuspendProfile($id: ID!) {
    suspendProfile(id: $id) {
      id
    }
  }
`;

export const UNSUSPEND_PROFILE = gql`
  mutation UnSuspendProfile($id: ID!) {
    unsuspendProfile(id: $id) {
      id
    }
  }
`;
