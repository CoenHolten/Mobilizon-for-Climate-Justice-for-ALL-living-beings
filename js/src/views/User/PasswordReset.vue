<template>
  <section class="columns is-mobile is-centered">
    <div class="card column is-half-desktop">
      <h1>
        <translate>Password reset</translate>
      </h1>
      <b-message title="Error" type="is-danger" v-for="error in errors" :key="error">{{ error }}</b-message>
      <form @submit="resetAction">
        <b-field label="Password">
          <b-input
            aria-required="true"
            required
            type="password"
            password-reveal
            minlength="6"
            v-model="credentials.password"
          />
        </b-field>
        <b-field label="Password (confirmation)">
          <b-input
            aria-required="true"
            required
            type="password"
            password-reveal
            minlength="6"
            v-model="credentials.password_confirmation"
          />
        </b-field>
        <button class="button is-primary">
          <translate>Reset my password</translate>
        </button>
      </form>
    </div>
  </section>
</template>

<script lang="ts">
import { Component, Prop, Vue } from "vue-property-decorator";
import { validateRequiredField } from "@/utils/validators";
import { RESET_PASSWORD } from "@/graphql/auth";
import { saveUserData } from "@/utils/auth";
import { ILogin } from "@/types/login.model";

@Component
export default class PasswordReset extends Vue {
  @Prop({ type: String, required: true }) token!: string;

  credentials = {
    password: "",
    password_confirmation: ""
  } as { password: string; password_confirmation: string };
  errors: string[] = [];
  rules = {
    password_length: value =>
      value.length > 6 || "Password must be at least 6 characters long",
    required: validateRequiredField,
    password_equal: value =>
      value === this.credentials.password || "Passwords must be the same"
  };

  get samePasswords() {
    return (
      this.rules.password_length(this.credentials.password) === true &&
      this.credentials.password === this.credentials.password_confirmation
    );
  }

  async resetAction(e) {
    e.preventDefault();
    this.errors.splice(0);

    try {
      const result = await this.$apollo.mutate<{ resetPassword: ILogin }>({
        mutation: RESET_PASSWORD,
        variables: {
          password: this.credentials.password,
          token: this.token
        }
      });

      saveUserData(result.data.resetPassword);
      this.$router.push({ name: "Home" });
    } catch (err) {
      console.error(err);
      err.graphQLErrors.forEach(({ message }) => {
        this.errors.push(message);
      });
    }
  }
}
</script>
