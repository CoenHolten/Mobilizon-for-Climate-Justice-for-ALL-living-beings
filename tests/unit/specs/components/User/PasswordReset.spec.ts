import { config, mount } from "@vue/test-utils";
import PasswordReset from "@/views/User/PasswordReset.vue";
import { createMockClient, RequestHandler } from "mock-apollo-client";
import { RESET_PASSWORD } from "@/graphql/auth";
import { resetPasswordResponseMock } from "../../mocks/auth";
import RouteName from "@/router/name";
import flushPromises from "flush-promises";
import { beforeEach, describe, expect, it, vi } from "vitest";
import { DefaultApolloClient } from "@vue/apollo-composable";
import Oruga from "@oruga-ui/oruga-next";
import {
  VueRouterMock,
  createRouterMock,
  injectRouterMock,
} from "vue-router-mock";

config.global.plugins.push(Oruga);
config.plugins.VueWrapper.install(VueRouterMock);

let requestHandlers: Record<string, RequestHandler>;

const generateWrapper = (
  customRequestHandlers: Record<string, RequestHandler> = {},
  customMocks = {}
) => {
  const mockClient = createMockClient();

  requestHandlers = {
    resetPasswordMutationHandler: vi
      .fn()
      .mockResolvedValue(resetPasswordResponseMock),
    ...customRequestHandlers,
  };

  mockClient.setRequestHandler(
    RESET_PASSWORD,
    requestHandlers.resetPasswordMutationHandler
  );

  return mount(PasswordReset, {
    props: {
      token: "some-token",
    },
    global: {
      stubs: ["router-link", "router-view"],
      mocks: {
        ...customMocks,
      },
      provide: {
        [DefaultApolloClient]: mockClient,
      },
    },
  });
};

describe("Reset page", () => {
  const router = createRouterMock({
    spy: {
      create: (fn) => vi.fn(fn),
      reset: (spy) => spy.mockReset(),
    },
  });
  beforeEach(() => {
    // inject it globally to ensure `useRoute()`, `$route`, etc work
    // properly and give you access to test specific functions
    injectRouterMock(router);
  });

  it("renders correctly", () => {
    const wrapper = generateWrapper();
    expect(wrapper.router).toBe(router);
    expect(wrapper.findAll('input[type="password"').length).toBe(2);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("shows error if token is invalid", async () => {
    const wrapper = generateWrapper({
      resetPasswordMutationHandler: vi.fn().mockResolvedValue({
        errors: [{ message: "The token you provided is invalid." }],
      }),
    });

    wrapper
      .findAll('input[type="password"')
      .forEach((inputField) => inputField.setValue("my password"));
    wrapper.find("form").trigger("submit");

    await wrapper.vm.$nextTick();

    expect(requestHandlers.resetPasswordMutationHandler).toBeCalledTimes(1);
    expect(requestHandlers.resetPasswordMutationHandler).toBeCalledWith({
      password: "my password",
      token: "some-token",
    });

    await flushPromises();

    expect(wrapper.find(".o-notification--danger").text()).toContain(
      "The token you provided is invalid"
    );
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("redirects to homepage if token is valid", async () => {
    const wrapper = generateWrapper();

    wrapper
      .findAll('input[type="password"')
      .forEach((inputField) => inputField.setValue("my password"));
    await wrapper.find("form").trigger("submit");

    expect(requestHandlers.resetPasswordMutationHandler).toBeCalledTimes(1);
    expect(requestHandlers.resetPasswordMutationHandler).toBeCalledWith({
      password: "my password",
      token: "some-token",
    });
    await flushPromises();
    expect(wrapper.router.push).toHaveBeenCalledWith({ name: RouteName.HOME });
  });
});
