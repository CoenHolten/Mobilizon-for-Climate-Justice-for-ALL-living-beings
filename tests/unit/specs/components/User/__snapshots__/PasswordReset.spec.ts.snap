// Vitest Snapshot v1

exports[`Reset page > renders correctly 1`] = `
"<section class=\\"container mx-auto\\">
  <h1 class=\\"\\">Password reset</h1>
  <form>
    <div class=\\"o-field\\"><label class=\\"o-field__label\\">Password</label>
      <div class=\\"o-ctrl-input\\"><input aria-required=\\"true\\" required=\\"\\" minlength=\\"6\\" class=\\"o-input o-input-iconspace-right\\" type=\\"password\\" autocomplete=\\"off\\">
        <!--v-if--><span class=\\"o-icon o-icon--clickable o-input__icon-right\\"><i class=\\"mdi mdi-eye mdi-24px\\"></i></span>
        <!--v-if-->
      </div>
      <!--v-if-->
    </div>
    <div class=\\"o-field\\"><label class=\\"o-field__label\\">Password (confirmation)</label>
      <div class=\\"o-ctrl-input\\"><input aria-required=\\"true\\" required=\\"\\" minlength=\\"6\\" class=\\"o-input o-input-iconspace-right\\" type=\\"password\\" autocomplete=\\"off\\">
        <!--v-if--><span class=\\"o-icon o-icon--clickable o-input__icon-right\\"><i class=\\"mdi mdi-eye mdi-24px\\"></i></span>
        <!--v-if-->
      </div>
      <!--v-if-->
    </div><button class=\\"button is-primary\\">Reset my password</button>
  </form>
</section>"
`;

exports[`Reset page > shows error if token is invalid 1`] = `
"<section class=\\"container mx-auto\\">
  <h1 class=\\"\\">Password reset</h1>
  <transition-stub title=\\"Error\\">
    <article class=\\"o-notification o-notification--danger\\">
      <!--v-if-->
      <!--v-if-->
      <div class=\\"o-notification__wrapper\\">
        <!--v-if-->
        <div class=\\"o-notification__content\\">The token you provided is invalid.</div>
      </div>
    </article>
  </transition-stub>
  <form>
    <div class=\\"o-field o-field--filled\\"><label class=\\"o-field__label\\">Password</label>
      <div class=\\"o-ctrl-input\\"><input aria-required=\\"true\\" required=\\"\\" minlength=\\"6\\" class=\\"o-input o-input-iconspace-right\\" type=\\"password\\" autocomplete=\\"off\\">
        <!--v-if--><span class=\\"o-icon o-icon--clickable o-input__icon-right\\"><i class=\\"mdi mdi-eye mdi-24px\\"></i></span>
        <!--v-if-->
      </div>
      <!--v-if-->
    </div>
    <div class=\\"o-field o-field--filled\\"><label class=\\"o-field__label\\">Password (confirmation)</label>
      <div class=\\"o-ctrl-input\\"><input aria-required=\\"true\\" required=\\"\\" minlength=\\"6\\" class=\\"o-input o-input-iconspace-right\\" type=\\"password\\" autocomplete=\\"off\\">
        <!--v-if--><span class=\\"o-icon o-icon--clickable o-input__icon-right\\"><i class=\\"mdi mdi-eye mdi-24px\\"></i></span>
        <!--v-if-->
      </div>
      <!--v-if-->
    </div><button class=\\"button is-primary\\">Reset my password</button>
  </form>
</section>"
`;
